### ISSUE TITLE
Excluding waveforms with constant components from Generator

### ISSUE THREAD

**Issue by liutianlin0121** [seisbench/seisbench#315]
Hi!

After reviewing a few datasets, such as NEIC, I noticed that some of the waveforms have constant channels. These constant channels have 0 std, and they make the std normalization unstable. Below are examples of these waveforms with constant-valued channels in NEIC:

![image](https://github.com/user-attachments/assets/185a6ed6-d8f3-44ed-89d3-35d8494b6e01)

These waveforms can be catched easily with:

```python
import seisbench.data as sbd
import matplotlib.pyplot as plt

data = sbd.NEIC()

NUM_TRACES_TO_COLLECT = 5

std_zero_waveforms = {}
for i in range(len(data)):
  waveforms = data.get_waveforms(i)
  std = waveforms.std(axis=1)
  if std.min() == 0:
    std_zero_waveforms[i] = waveforms
    if len(std_zero_waveforms) > NUM_TRACES_TO_COLLECT:
      break

for idx, waveforms in std_zero_waveforms.items():
  plt.figure(figsize=(5, 2))
  plt.plot(waveforms.T)
  plt.title(f"Trace {idx} in NEIC")
  plt.show()
``` 

My question is: Is there a way to exclude these waveforms with constant channels from the Generator? 

To solve this, my first attempt was to check if the meta data recorded these waveforms with constant channels, which would allow me to use SteeredGenerator. However, it seems that the meta data, at least for NEIC, does not include this information.

I think that another way is to do some kind of rejecting sampling in the augmentation module (if a waveform has constant channels, then exclude it). I looked into the [sbg code](https://seisbench.readthedocs.io/en/stable/_modules/seisbench/generate/generator.html#GenericGenerator), but I couldn’t find anything related to this functionality.

Do you have any suggestions? Maybe there are easier ways to address this? I appreciate your help!

**Comment by yetinam** [comment]
Hi @liutianlin0121 ,
thanks for pointing this out! First thing to check is the `trace_component_order` in the metadata. Best case, the traces with zeros will have fewer components listed here (in which case the zeros are only added by SeisBench when loading). I. This case, the filter function could help you,

However, not all datasets use his possibility. To filter the zeros in loading, the easiest option is probably to subclass the GenericGenerator and overwrite the `_populate_state_dict` function.

If you'd like, you can also submit a PR to SeisBench to add this functionality as an option to the GenericGenerator. I'd be happy to include it in SeisBench.

**Comment by liutianlin0121** [comment]
Thank you for your response!

With more checking, I think these constant channels mainly occur in the NEIC dataset. Also, the fraction of the samples with constant channels is only a few percent. So at this moment, I slightly modified the normalization module, which keeps 0-std waveforms unchanged when std norm is used. This allows me to use std normalization without numerical instability: 

```python
class StdSafeNormalize(sbg.Normalize):

  def _amp_norm(self, x: np.ndarray) -> np.ndarray:
    if self.amp_norm_axis is not None:
      if self.amp_norm_type == "peak":
        x = x / (
            np.max(np.abs(x), axis=self.amp_norm_axis, keepdims=True) + self.eps
        )
      elif self.amp_norm_type == "std":
        std = np.std(x, axis=self.amp_norm_axis, keepdims=True)
        std = np.where(std == 0, 1, std) # set zero-std waveforms to zero. 
        x = x / (std + self.eps)
    return x
```
